src/coreclr/interpreter/compiler.cpp

```diff
@@ -0,0 +1,583 @@
+        pBB->pFirstIns = ins;
+    }
+
+    if (ins->pNext == NULL)
+    {
+        pBB->pLastIns = ins;
+    }
+    else
+    {
+        ins->pNext->pPrev = ins;
+    }
+
+    return ins;
+}
+
+// Inserts a new instruction after prevIns. prevIns must be in cbb
+InterpInst* InterpCompiler::InsertIns(InterpInst *pPrevIns, int opcode)
+{
+    return InsertInsBB(m_pCBB, pPrevIns, opcode);
+}
+
+InterpInst* InterpCompiler::FirstRealIns(InterpBasicBlock *pBB)
+{
+    InterpInst *ins = pBB->pFirstIns;
+    if (!ins || !InsIsNop(ins))
+        return ins;
+    while (ins && InsIsNop(ins))
+        ins = ins->pNext;
+    return ins;
+}
+
+InterpInst* InterpCompiler::NextRealIns(InterpInst *ins)
+{
+    ins = ins->pNext;
+    while (ins && InsIsNop(ins))
+        ins = ins->pNext;
+    return ins;
+}
+
+InterpInst* InterpCompiler::PrevRealIns(InterpInst *ins)
+{
+    ins = ins->pPrev;
+    while (ins && InsIsNop(ins))
+        ins = ins->pPrev;
+    return ins;
+}
+
+void InterpCompiler::ClearIns(InterpInst *ins)
+{
+    ins->opcode = INTOP_NOP;
+}
+
+bool InterpCompiler::InsIsNop(InterpInst *ins)
+{
+    return ins->opcode == INTOP_NOP;
+}
+
+int32_t InterpCompiler::GetInsLength(InterpInst *ins)
+{
+    return g_interpOpLen[ins->opcode];
+}
+
+void InterpCompiler::ForEachInsSVar(InterpInst *ins, void *pData, void (InterpCompiler::*callback)(int*, void*))
+{
+    int numSVars = g_interpOpSVars[ins->opcode];
+    if (numSVars)
+    {
+        for (int i = 0; i < numSVars; i++)
+        {
+            if (ins->sVars [i] == CALL_ARGS_SVAR)
+            {
+                if (ins->info.pCallInfo && ins->info.pCallInfo->pCallArgs) {
+                    int *callArgs = ins->info.pCallInfo->pCallArgs;
+                    while (*callArgs != -1)
+                    {
+                        (this->*callback) (callArgs, pData);
+                        callArgs++;
+                    }
+                }
+            }
+            else
+            {
+                (this->*callback) (&ins->sVars[i], pData);
+            }
+        }
+    }
+}
+
+void InterpCompiler::ForEachInsVar(InterpInst *ins, void *pData, void (InterpCompiler::*callback)(int*, void*))
+{
+    ForEachInsSVar(ins, pData, callback);
+
+    if (g_interpOpDVars [ins->opcode])
+        (this->*callback) (&ins->dVar, pData);
+}
+
+
+InterpBasicBlock* InterpCompiler::AllocBB()
+{
+    InterpBasicBlock *bb = (InterpBasicBlock*)AllocMemPool(sizeof(InterpBasicBlock));
+    memset(bb, 0, sizeof(InterpBasicBlock));
+    bb->ilOffset = -1;
+    bb->nativeOffset = -1;
+    bb->stackHeight = -1;
+    bb->index = m_BBCount++;
+    return bb;
+}
+
+InterpBasicBlock* InterpCompiler::GetBB(int32_t ilOffset)
+{
+    InterpBasicBlock *bb = m_ppOffsetToBB [ilOffset];
+
+    if (!bb)
+    {
+        bb = AllocBB ();
+
+        bb->ilOffset = ilOffset;
+        m_ppOffsetToBB[ilOffset] = bb;
+    }
+
+    return bb;
+}
+
+// FIXME Replace this
+static inline uint32_t ClzI4(uint32_t val)
+{
```

Comment:
There is a `__builtin_clz` provided by Unix compilers and `_BitScanReverse` by MSVC that you can use. JIT uses them too (see BitOperations::BitScanReverse)